cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,336 @@
+                # Add the newly generated cell to the list of living lineages
+                current_lineages.append(
+                    {
+                        "id": unique_id,
+                        "birth_scale": lineage["birth_scale"]
+                        * total_birth_mutation_strength,
+                        "total_time": birth_waiting_time
+                        + lineage["total_time"],
+                    }
+                )
+                return unique_id + 1
+            else:
+                return unique_id
+
+        # Instantiate the implicit root
+        tree = nx.DiGraph()
+        tree.add_node(0)
+        tree.nodes[0]["birth_scale"] = birth_scale_param
+        tree.nodes[0]["total_time"] = 0
+        current_lineages = []
+        starting_lineage = {
+            "id": 0,
+            "birth_scale": birth_scale_param,
+            "total_time": 0,
+        }
+
+        # Sample the waiting time until the first division
+        unique_id = 1
+        unique_id = sample_event(unique_id, starting_lineage)
+
+        # Perform the process until there are no active extant lineages left
+        while len(current_lineages) > 0:
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if num_extant:
+                if len(current_lineages) == num_extant:
+                    min_total_time = min(
+                        [i["total_time"] for i in current_lineages]
+                    )
+                    for remaining_lineage in current_lineages:
+                        parent = list(
+                            tree.predecessors(remaining_lineage["id"])
+                        )[0]
+                        tree.edges[parent, remaining_lineage["id"]][
+                            "weight"
+                        ] += (min_total_time - remaining_lineage["total_time"])
+                        tree.nodes[remaining_lineage["id"]][
+                            "birth_scale"
+                        ] = tree.nodes[parent]["birth_scale"]
+                    break
+            # If extant tips are the stopping criteria, pop the minimum age
+            # lineage at each step
+            if num_extant:
+                min_age_ind = np.argmin(
+                    [i["total_time"] for i in current_lineages]
+                )
+                lineage = current_lineages.pop(min_age_ind)
+            else:
+                lineage = current_lineages.pop(0)
+            for _ in range(2):
+                unique_id = sample_event(unique_id, lineage)
+
+        # Prune dead lineages and collapse resulting unifurcations
+        if death_waiting_dist and len(tree.nodes) > 1:
```

Comment:
it shouldn't matter whether or not `death_waiting_dist` is provided, right? If it is not provided, the code below simply won't have anything to prune, but should still run?